src/modules/roc_netio/target_uv/roc_netio/transceiver.cpp

```diff
@@ -214,12 +198,88 @@ void Transceiver::stop_sem_cb_(uv_async_t* handle) {
     roc_panic_if_not(handle);
 
     Transceiver& self = *(Transceiver*)handle->data;
-    self.stop_();
+    self.stop_all_();
     self.close_();
     self.process_tasks_();
 }
 
+void Transceiver::close_cb_(uv_handle_t* handle) {
+    roc_panic_if_not(handle);
+
+    Transceiver& self = *(Transceiver*)handle->data;
+
+    core::Mutex::Lock lock(self.mutex_);
+
+    if (handle == (uv_handle_t*)&self.stop_sem_) {
+        self.stop_sem_initialized_ = false;
+    } else {
+        self.task_sem_initialized_ = false;
+    }
+
+    if (self.stop_sem_initialized_ || self.task_sem_initialized_) {
+        return;
+    }
+
+    self.cond_.broadcast();
+}
+
+void Transceiver::remove_port_cb_(void* data, packet::Address& address) {
+    roc_log(LogDebug, "transceiver: removed port %s",
+            packet::address_to_str(address).c_str());
+
+    Transceiver& self = *(Transceiver*)data;
+
+    core::Mutex::Lock lock(self.mutex_);
+
+    core::SharedPtr<UDPReceiver> rp = self.get_receiver_(address);
+    if (!rp) {
+        core::SharedPtr<UDPSender> sp = self.get_sender_(address);
+        if (!sp) {
+            roc_panic("transceiver: can't remove unknown port %s",
+                      packet::address_to_str(address).c_str());
+        } else {
+            self.senders_.remove(*sp);
+        }
+    } else {
+        self.receivers_.remove(*rp);
```

Comment:
The nesting is a bit confusing. It would be cleaner to extract a new method that uses the same nesting for consecutive operations:

```
    core::SharedPtr<UDPReceiver> rp = self.get_receiver_(address);
    if (rp) {
        self.senders_.remove(*sp);
        return;
    }

    core::SharedPtr<UDPSender> sp = self.get_sender_(address);
    if (sp) {
        self.receivers_.remove(*sp);
        return;
    }

    roc_panic(...);
```

Reply:
However, there is a better option. Current approach has one problem: we should add code for every new element type (e.g. when we'll add TCPSender and TCPReceiver) here and in a few other places. Instead, we can do the following:

* add a new interface, say, `IElement`
* make UDPReceiver and UDPSender to inherit this interface
* use a single list for senders and receivers and unify the removing code

Something like this:

```
class IElement : public RefCnt<IElement> {
public:
    virtual void stop() = 0;
};
```

This will also allow to use `IElement` instead of `void*` in `ICloseHandler`.

Reply:
done